keps/sig-network/0031-service-topology.md

```diff
@@ -0,0 +1,218 @@
+
+Endpoints controller will need to watch Nodes for knowing labels of node hosting the endpoint and copy the node labels referenced in the service spec's topology constraints to EndpointAddress.
+
+Endpoints Controller will also maintain an extra cache: `NodeToPodsCache`.
+`NodeToPodsCache` maps the node's name to the pods running on it. Node's add, delete and labels' change will trigger `NodeToPodsCache` re-index.
+
+So, the new logic of endpoint controller might be like:
+
+```go
+go watch Node
+// In each sync loop, for a given service, sync its endpoints
+for i, pod := range service backends; do
+  node := nodeCache[pod.Spec.NodeName]
+  endpointAddress := &v1.EndpointAddress {}
+  // Copy all topology-related labels of node to all the endpoints running on it.
+  // We can only include node labels referenced in the service spec's topology constraints
+  for _, topologyKey := range service.TopologyKeys; do
+    endpointAddress.Topologies[topoKey] = node.Labels[topologyKey]
+  done
+  endpoints.Subsets[i].Addresses = endpointAddress
+done
+```
+
+### Kube-proxy changes
+
+Kube-proxy will respect topology keys for each service, so kube-proxy on different nodes may create different proxy rules.
+
+Kube-proxy will watch or periodically get(which approach has better performance?) its own node and will find the endpoints that are in the same topological domain as the node if `service.TopologyKeys` is not empty.
+
+The new logic of kube-proxy might be like:
+
+```go
+go watch/periodically get node with its nodename
+endpointsMeetRequirement := make([]endpointInfo, 0)
+for _, topologyKey := range service.TopologyKeys; do
+  for i := range service's endpoints.Subsets; do
+    ss := endpoints.Subsets[i]
+    for j := range ss.Addresses; do
+      // check if endpoint are in the same topological domain as the node running kube-proxy
+      if ss.Addresses[j].TopologyKey[topologyKey] == node.Labels[topologyKey]; then
+        endpointsMeetRequirement = append(endpointsMeetHardRequirement, endpoint)
+      fi
+    done
+    // Randomly pick one if there are some endpoints(>=1) matched
+    if len(endpointsMeetRequirement) != 0; then
+      route request to an endpoint randomly
+      return
+    fi
+  done
+done
+connection fails due to no mactch endpoint
+```
+
+### DNS changes
+
+We should consider this kind of topology support for headless service in coredns and kube-dns. As the DNS servers will respect topology keys for each headless service, different clients/pods on different nodes may get different dns response.
+
+#### CoreDNS changes
+
+CoreDNS already has the client IP to Pod mapping and it will need to watch nodes to get topology-related information. When a client/pod request a headless service domain to CoreDNS, CoreDNS will retrieve the node labels of both client and the backend Pods. CoreDNS will only select the IPs of backend Pods which are in the same topological domain with client Pod, and then write A record.
```

Comment:
How does CoreDNS track IP->Pod?  Is it watching all pods?  That is a dangerous road...

@johnbelamaric

Reply:
@thockin, 

CoreDNS does not do this by default.  But yes, watching Pods is an option.  For example, it's used by the `autopath` plugin for determining the namespace of the client Pod from the Pod IP.

> That is a dangerous road.
 
Other than the extra memory used by the Pod store, what other concerns are there?
Significantly more API watch overhead?
Other dangers?

Reply:
The load generated by watching all pods is significant.  DNS tends to be very scale-out in a lot of clusters, so having many more all-pod watches can be disastrous.

Reply:
I fully agree with Tim. In very large clusters, I can easily imagine hundreds of DNS instances running. And having hundreds of "all pods watchers" doesn't seem to be desired solution.
In a way, we would be repeating the problematic pattern of kube-proxy, where each kube-proxy is watching all endpoints - that is a real painpoint from scalability/performance perspective.

Reply:
So I assume that's the mission of SIG scalability.

And do you have any constructive suggestion on better server-side field-filtering as Tim commented?

Reply:
So, our choices with respect to headless services that I can think of are:
1) Cluster DNS servers watch all pods to do source IP -> node mapping and apply service policy.
2) A node-local proxy (DNS caching or just mutating proxy) appends node name to the outbound request so the central DNS servers can apply the mapping.
3) A node-local DNS cache filters the responses from the upstream (seems like a very bad idea)
4) A node-local DNS cache that is pushed all headless services and weeds out the IPs based on local node topology (similar to 3 but slightly different mechanism, and probably an equally bad idea)
5) Build a new object type that just contains Pod IP -> Node mapping (or Pod IP -> Node topologies), and watch that instead of all pods (new api object which we are trying to avoid)
6) Implement some sort of watch selection mechanism that allows us to only watch for changes on a pod to IP or node (or equivalently in this case that allows us to select only create/delete events), assuming that reduces load on the apiserver (big assumption) 
7) Some horribly complicated thing where a few of the cluster DNS servers are allowed to watch pods and then they distribute the necessary info to the other cluster DNS servers
8) Not support headless services for this feature by default (users with small clusters can enable it in CoreDNS options if they want)
9) Not support headless services for this feature at all (reject service API create/update that set a topology restriction on headless services)

In my opinion, if we can assume #2758 then I would prefer option 2. Without that I prefer option 8 (which implies option 1 but only for clusters that explicitly enable it).

Reply:
I think we can start with:

> 8. Not support headless services for this feature by default (users with small clusters can enable it in CoreDNS options if they want)

And consider:

> 5. Build a new object type that just contains Pod IP -> Node mapping (or Pod IP -> Node topologies), and watch that instead of all pods (new api object which we are trying to avoid)

Having a controller  that watches pods and writes a new resource which only carries the pod IP(s) and the node name doesn't seem horrible.  Adding a new resource is not ideal, but we can either consider this a DNS-specific resource/controller or make an argument for doing other things with it (e.g. reverse DNS - @bowei  @pavithrar).  If we had this, what other problems get easier?

@caseydavenport Does this make Typha easier (or what would be needed to do so) ?

@wojtek-t what would scale team think about this? :)